packages/sanity/src/core/store/events/getDocumentEvents.ts

```diff
@@ -0,0 +1,320 @@
+function getEventFromTransaction(
+  documentId: string,
+  transaction: Transaction,
+  previousTransactions: Transaction[],
+): DocumentGroupEvent | null {
+  const base = {
+    timestamp: transaction.timestamp,
+    author: transaction.author,
+  }
+
+  const draftId = getDraftId(documentId)
+  const publishedId = getPublishedId(documentId)
+  const draftEffect = transaction.effects[draftId]
+  const publishedEffect = transaction.effects[publishedId]
+  const draftState = getEffectState(draftEffect)
+  const publishedState = getEffectState(publishedEffect)
+
+  const getDocumentEvent = (
+    type: DocumentEventType,
+    document: DocumentToMapTheAction,
+  ): DocumentGroupEvent | null => {
+    switch (type) {
+      case 'document.createVersion': {
+        return {
+          ...base,
+          type,
+
+          documentId: documentId,
+          versionId: document === 'draft' ? draftId : publishedId,
+          releaseId: getVersionFromId(documentId),
+          versionRevisionId: transaction.id,
+        }
+      }
+      case 'document.editVersion': {
+        return {
+          ...base,
+          type,
+
+          releaseId: getVersionFromId(documentId),
+          versionId: draftId,
+          versionRevisionId: transaction.id,
+        }
+      }
+
+      case 'document.deleteVersion': {
+        return {
+          ...base,
+          type,
+
+          versionId: document === 'draft' ? draftId : publishedId,
+          // The revision id of the last edit in the draft document
+          versionRevisionId:
+            previousTransactions.find((t) => t.documentIDs.includes(draftId))?.id || 'not-found',
+          releaseId: getVersionFromId(documentId),
+        }
+      }
+
+      case 'document.publishVersion': {
+        return {
+          ...base,
+          type,
+
+          revisionId: transaction.id,
+          releaseId: getVersionFromId(documentId),
+          versionId: draftId, // TODO: How to get the version in case of releases
+          versionRevisionId:
+            previousTransactions.find((t) => t.documentIDs.includes(draftId))?.id || 'not-found',
+          cause: {
+            type: 'document.publish', // TODO: How to get the `release.publish` and the `release.schedule` events?
+          },
+        }
+      }
+
+      case 'document.unpublish': {
+        return {
+          ...base,
+          type,
+
+          // The version that will be created by this unpublish action, e.g. drafts.foo
+          versionId:
+            transaction.documentIDs.length > 1 && transaction.documentIDs.includes(draftId)
+              ? draftId
+              : undefined,
+          // TODO: Should this be the actual transaction id that created the draft document?
+          // The revision id of the last edit in the publish document
+          versionRevisionId:
+            previousTransactions.find((t) => t.documentIDs.includes(publishedId))?.id ||
+            'not-found',
```

Comment:
I'm unsure on what we want to express here:
 - the last time the published id was modified (last revision id for the published document)
 - revisionId of the draft document that was created by this transaction, that would be the `transaction.id`

Reply:
I think we want the second option.

Reply:
Yes, makes more sense like that